File: de.unistuttgart.iste.sqa.mpw.hamstersimulator.java/core/src/test/java/de/unistuttgart/hamster/util/GameStringifier.java
package de.unistuttgart.hamster.util;

import de.unistuttgart.hamster.hamster.*;
import de.unistuttgart.hamster.mpw.Tile;

import java.util.function.Consumer;

public class GameStringifier {
	
	public static HamsterGame createFromString(String map) {
		var game = new HamsterGame();

		String[] parts = map.split(";");
		int height = parts.length;
		int width = height > 0 ? parts[0].length() : 0;

		var territoryBuilder = new TerritoryBuilder(game);
		territoryBuilder.initTerritory(width, height);

		for (int y = 0; y < height; y++) {
			handleLine(territoryBuilder, y, parts[y]);
		}

		game.startGame();
		game.getPerformance().setDelayEnabled(false);

		return game;
	}

	private static void handleLine(TerritoryBuilder territoryBuilder, int y, String part) {
		char[] chars = part.toCharArray();
		for (int x = 0; x < part.length(); x++) {
			handleCell(territoryBuilder, x, y, chars[x]);
		}
	}

	private static void handleCell(TerritoryBuilder territoryBuilder, int x, int y, char cell) {
		switch (cell) {
			case '>', '<', '^', 'v' -> territoryBuilder.initDefaultHamster(x, y, DirectionTestHelper.toDirection(cell), 0);
			case '*' -> territoryBuilder.addGrainsToTile(x, y, 1);
			case 'M' -> territoryBuilder.addWallToTile(x, y);
		}
	}

	public static String toString(HamsterGame game) {
		var actual = new StringBuilder();

		var hamster = game.getTerritory().getDefaultHamster().getInternalHamster();
		iterateTiles(game, currentTile -> {
			if (hamster.getCurrentTile() == currentTile) {
				actual.append(DirectionTestHelper.toDirection(hamster.getDirection()));
			} else if (currentTile.getContents().stream().anyMatch(Wall.class::isInstance)) {
				actual.append('M');
			} else if (currentTile.getContents().stream().anyMatch(Grain.class::isInstance)) {
				actual.append('*');
			} else {
				actual.append(' ');
			}

			if (currentTile.getEast() == null) {
				actual.append(';');
			}
		});

		return actual.toString();
	}

	public static String grainsOnTerritoryToString(HamsterGame game) {
		var actual = new StringBuilder();

		iterateTiles(game, currentTile -> {
			var grainsCount = currentTile.getContents().stream().filter(Grain.class::isInstance).count();
			actual.append(grainsCount);

			if (currentTile.getEast() == null) {
				actual.append(';');
			}
		});

		return actual.toString();
	}

	private static void iterateTiles(HamsterGame game, Consumer<Tile> lambda) {
		var tiles = game.getTerritory().getInternalTerritory().getTiles();
		final var upperLeftTile = tiles.get(0);
		var firstOfRowTile = upperLeftTile;

		var currentTile = upperLeftTile;

		while (currentTile != null) {
			lambda.accept(currentTile);

			if (currentTile.getEast() == null) {
				firstOfRowTile = firstOfRowTile.getSouth();
				currentTile = firstOfRowTile;
			} else {
				currentTile = currentTile.getEast();
			}
		}
	}

}

File: de.unistuttgart.iste.sqa.mpw.hamstersimulator.cpp/core/test/util/GameStringifier.cpp
#include <CollectionHelpers.hpp>
#include "GameStringifier.h"

#include "Tile.h"
#include "Grain.h"
#include "Wall.h"
#include "Territory.h"
#include "Hamster.h"
#include "GameTerritory.h"
#include "GameHamster.h"
#include "GamePerformance.h"

#include "TerritoryBuilder.h"
#include "DirectionTestHelper.h"
#include "TestUtils.h"

using namespace mpw;
using namespace hamster;

namespace util {

static void handleLine(TerritoryBuilder& territoryBuilder, int y, const std::string& part);
static void handleCell(TerritoryBuilder& territoryBuilder, int x, int y, char cell);

std::shared_ptr<hamster::HamsterGame> GameStringifier::createFromString(const std::string& map) {
    auto game = HamsterGame::create();

    std::vector<std::string> parts = TestUtils::split(";", map);
    int height = static_cast<int>(parts.size());
    int width = height > 0 ? static_cast<int>(parts[0].length()) : 0;

    auto territoryBuilder = std::make_shared<TerritoryBuilder>(game);
    territoryBuilder->initTerritory(width, height);

    for (int y = 0; y < height; y++) {
        handleLine(*territoryBuilder, y, parts[y]);
    }

    game->getPerformance()->setDelayEnabled(false);
    game->startGame();
    return game;
}

static void handleLine(TerritoryBuilder& territoryBuilder, int y, const std::string& part) {
    for (int x = 0; x < part.length(); x++) {
        handleCell(territoryBuilder, x, y, part[x]);
    }
}

static void handleCell(TerritoryBuilder& territoryBuilder, int x, int y, char cell) {
    switch (cell) {
        case '>':
        case '<':
        case '^':
        case 'v':
            territoryBuilder.initDefaultHamster(x, y, DirectionTestHelper::toDirection(cell), 0);
            break;
        case '*':
            territoryBuilder.addGrainsToTile(x, y, 1);
            break;
        case 'M':
            territoryBuilder.addWallToTile(x, y);
            break;
        default: ;
    }
}

static void iterateTiles(const HamsterGame& game, const std::function<void(Tile&)>& lambda);

std::string GameStringifier::toString(const HamsterGame& game) {
    std::string actual;

    auto& nonConstGame = const_cast<HamsterGame&>(game); // TODO fix this const-correctness issue
    auto hamster = nonConstGame.getTerritory()->getDefaultHamster()->getInternalHamster();
    iterateTiles(game, [&](Tile& currentTile) {
        if (hamster->getCurrentTile().get() == &currentTile) {
            actual += util::DirectionTestHelper::toDirection(hamster->getDirection());
        } else if (!collectionhelpers::type_select<Wall>(currentTile.getContents()).empty()) {
            actual += 'M';
        } else if (!collectionhelpers::type_select<Grain>(currentTile.getContents()).empty()) {
            actual += '*';
        } else {
            actual += ' ';
        }

        if (currentTile.getEast() == nullptr) {
            actual += ';';
        }
    });

    return actual;
}

std::string GameStringifier::grainsOnTerritoryToString(const HamsterGame& game) {
    std::string actual;

    iterateTiles(game, [&](Tile& currentTile) {
        auto grainsCount = collectionhelpers::type_select<Grain>(currentTile.getContents()).size();
        actual += std::to_string(grainsCount);

        if (currentTile.getEast() == nullptr) {
            actual += ';';
        }
    });

    return actual;
}

static void iterateTiles(const HamsterGame& game, const std::function<void(Tile&)>& lambda) {
    auto& nonConstGame = const_cast<HamsterGame&>(game); // TODO fix this const-correctness issue
    auto tiles = nonConstGame.getTerritory()->getInternalTerritory()->getTiles();
    auto upperLeftTile = collectionhelpers::get_at(tiles, 0).value();
    auto firstOfRowTile = upperLeftTile;

    std::shared_ptr<Tile> currentTile = upperLeftTile;

    while (currentTile != nullptr) {
        lambda(*currentTile);

        if (currentTile->getEast() == nullptr) {
            firstOfRowTile = firstOfRowTile->getSouth();
            currentTile = firstOfRowTile;
        } else {
            currentTile = currentTile->getEast();
        }
    }
}

}
